### PR TITLE
add a schema introspection task

### LINKED ISSUE #91: Add a schema download task
apollo-codegen supports downloading the schema by sending an introspection query to the server.
It would be nice to have a task that utilizes `apollo-codegen download`

### PR DESCRIPTION
**Pull request by marwanad**
This task uses apollo-codegen to introspect a schema from a GraphQL server using `apollo-codegen introspect-schema`.

Usage, add the following to the build file:

```
task introspectSchema(type: com.apollographql.android.gradle.ApolloSchemaIntrospectionTask) {
  url = "http://api.githunt.com/graphql"
  output = "src/main/graphql/schema.json"
}
```
And run `gradle introspectSchema`

Closes #91.

### CODE REVIEW

File: apollo-gradle-plugin/src/main/java/com/apollographql/android/gradle/ApolloSchemaIntrospectionTask.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  public void setUrl(String url) {
+    this.url = url;
+  }
+
+  public void setOutput(String output) {
+    this.output = output;
+  }
+
+  public void setHeader(String header) {
```
Comment (sav007) [inline apollo-gradle-plugin/src/main/java/com/apollographql/android/gradle/ApolloSchemaIntrospectionTask.java]: Only one header can be set? What if I have to set multiple

**Comment by sav007** [comment]
`gradle introspectSchema` can I run this task with arguments from command line?

File: apollo-gradle-plugin/src/main/java/com/apollographql/android/gradle/ApolloSchemaIntrospectionTask.java (hunk 1 above)
Comment (marwanad) [inline apollo-gradle-plugin/src/main/java/com/apollographql/android/gradle/ApolloSchemaIntrospectionTask.java]: will change to a list/varargs.

**Comment by marwanad** [comment]
Not currently but can do that instead. It would still require defining the task in your build file though.

**Comment by sav007** [comment]
@marwanad any updates on this PR?

**Comment by marwanad** [comment]
@sav007 will get back to it by tomorrow. It can go with v0.3.2 instead.

**Comment by marwanad** [comment]
Updated. Arguments can be passed through command line now. 

Define the task inside you the build file
```
task introspectSchema(type: com.apollographql.android.gradle.ApolloSchemaIntrospectionTask)
```
and then
```
gradle introspectSchema --url "http://api.githunt.com/graphql" --output 
"src/main/graphql/schema.json"
```

**Comment by digitalbuddha** [comment]
Is a user able to define the params within gradle file as well?

On May 15, 2017 12:28 AM, "Marwan Ahmed" <notifications@github.com> wrote:

Updated. Arguments can be passed through command line now.

Define the task inside you the build file

task introspectSchema(type:
com.apollographql.android.gradle.ApolloSchemaIntrospectionTask)

and then

gradle introspectSchema --url "http://api.githunt.com/graphql" --output
"src/main/graphql/schema.json"

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/apollographql/apollo-android/pull/461#issuecomment-301341238>,
or mute the thread

**Comment by marwanad** [comment]
Yup, doing that works too:

```
task introspectSchema(type: com.apollographql.android.gradle.ApolloSchemaIntrospectionTask) {
  url = "http://api.githunt.com/graphql"
  output = "src/main/graphql/schema.json"
}
```

**Pull request merged**